### ISSUE TITLE
Performance Issues with STAsBinary

### ISSUE THREAD

**Issue by BHare1985** [gaspardle/mapnik-mssql#3]
Using the OGR vs your plugin seems to do things a bit differently and yours is slower.

First thing to note is that using `layer_by_sql`in for MSSQL is broke for OGR in mapnik because it won't apply the WHERE clause to intersect ( see: https://github.com/jbrwn/NET-Mapnik/issues/18#issuecomment-104394213 ) . So I am forced to use the `table` parameter which selects all fields, even then OGR is faster though.

With OGR the same tile request SQL looks like:

```
select [id], [point], [lng], [lat] from dbo.points where [point].STIntersects(geometry::STGeomFromText('POLYGON((-106.740281 25.86604,-98.4375 25.86604,-98.4375 34.307143856288,-106.740281 34.307143856288,-106.740281 25.86604))',4326)) = 1
```

Compared to mapnik-plugin:

```
SELECT "point".STAsBinary() AS geom FROM (SELECT point FROM points) as data WHERE "point".STIsValid() = 1 AND "point".STIntersects(geometry::STLineFromText('LINESTRING(-106.740281 25.86604,-98.43749999999999 34.30714385628804)', 4326).STEnvelope()) = 1
```

I understand you are doing things differently mostly for a purpose, but your query takes almost twice the time when getting 45814 rows ( Profiler proof: http://i.imgur.com/nTYaHUc.png)

I figured this was related to doing a subquery in the `table` and decided to just use `<Parameter name="table"><![CDATA[points]]></Parameter>`

This gave me the SQL: 

```
SELECT "point".STAsBinary() AS geom FROM points WHERE "point".STIsValid() = 1 AND "point".STIntersects(geometry::STLineFromText('LINESTRING(-106.740281 25.86604,-98.43749999999999 34.30714385628804)', 4326).STEnvelope()) = 1
```

The result was no different, taking around ~3700 ms.
Changing it to `SELECT "point" FROM points...` got it down to ~2500 ms
Getting rid of `"point".STIsValid() = 1 AND` got it down to ~1800ms comparable to OGR.

I am not sure why STAsBinary() and these checks are being made, but it seems my data was already in a valid format and don't require the checks. 

datasource for OGR:

```
    <Datasource>
       <Parameter name="type">ogr</Parameter>
       <Parameter name="string">MSSQL:Server=C-256\sqlexpress;Database=points;Uid=sa;Pwd=708050;Tables=points</Parameter>
       <Parameter name="layer">dbo.points</Parameter>
    </Datasource>
```

I am using the following datasource for mapnik-mssql plugin:

```
    <Datasource>       
       <Parameter name="type"><![CDATA[mssql]]></Parameter>   
       <Parameter name="host"><![CDATA[C-256\sqlexpress]]></Parameter>
       <Parameter name="driver"><![CDATA[SQL Server Native Client 11.0]]></Parameter>
       <Parameter name="dbname"><![CDATA[points]]></Parameter>
       <Parameter name="user"><![CDATA[sa]]></Parameter>
       <Parameter name="password"><![CDATA[pass]]></Parameter>
        <Parameter name="srid"><![CDATA[4326]]></Parameter>     
       <Parameter name="table"><![CDATA[(SELECT point FROM points) as data]]></Parameter>
       <Parameter name="geometry_field"><![CDATA[point]]></Parameter>
       <Parameter name="extent_cache"><![CDATA[auto]]></Parameter> 
    </Datasource>
```

**Comment by BHare1985** [comment]
I just noticed the wkb parameter in the new wiki, felt dumb. Turned it to false and tried again. Strangely it still uses STAsBinary with wkb = false.

**Comment by BHare1985** [comment]
On the issue of performance, I also noticed the

SELECT ext.STPointN(1).STX AS MinX, ext.STPointN(1).STY AS MinY,ext.STPointN(3).STX AS MaxX, ext.STPointN(3).STY AS MaxX FROM (SELECT geometry::EnvelopeAggregate(point) as ext from points) as tmp

Is being ran 4 times, and each query seems to have it's own Audit Login and Audit Logout. I assume this is related to turning async on by default.

**Comment by gaspardle** [comment]
Hi, there still some work to do in regard of performance.
The query uses .STIsValid() because if the table contains an invalid geometry, the whole query will fail when converting from the SQL Server CLR type to wkb.
The wkb parameter will fix this issue, but it is not yet in the latest binary that I published.

For the query running 4 times , you can use the extent parameter.
ex.:
`<Parameter name="extent"><![CDATA[-20037508.34,-20037508.34,20037508.34,20037508.34]]></Parameter>`

**Comment by BHare1985** [comment]
This is fixed in release v0.1.5-alpha. Thanks!
